File: python/basic/image_matching.py
"""
图像匹配相关问题：
56. 模板匹配(SSD) - 使用平方差和进行模板匹配
57. 模板匹配(SAD) - 使用绝对差和进行模板匹配
58. 模板匹配(NCC) - 使用归一化互相关进行模板匹配
59. 模板匹配(ZNCC) - 使用零均值归一化互相关进行模板匹配
60. 特征点匹配 - 使用特征描述子进行图像匹配
"""

import cv2
import numpy as np

def ssd_matching(img_path, template_path):
    """
    问题56：模板匹配(SSD)
    使用平方差和进行模板匹配

    参数:
        img_path: 输入图像路径
        template_path: 模板图像路径

    返回:
        匹配结果可视化
    """
    # 读取图像
    img = cv2.imread(img_path, cv2.IMREAD_GRAYSCALE)
    template = cv2.imread(template_path, cv2.IMREAD_GRAYSCALE)
    if img is None or template is None:
        raise ValueError("无法读取图像")

    h, w = template.shape
    H, W = img.shape
    result = np.zeros((H-h+1, W-w+1), dtype=np.float32)

    # 计算SSD
    for y in range(H-h+1):
        for x in range(W-w+1):
            diff = img[y:y+h, x:x+w] - template
            result[y, x] = np.sum(diff * diff)

    # 归一化结果
    result = cv2.normalize(result, None, 0, 255, cv2.NORM_MINMAX).astype(np.uint8)

    # 找到最佳匹配位置
    min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(result)
    top_left = min_loc
    bottom_right = (top_left[0] + w, top_left[1] + h)

    # 在原图上绘制矩形框
    img_color = cv2.cvtColor(img, cv2.COLOR_GRAY2BGR)
    cv2.rectangle(img_color, top_left, bottom_right, (0, 0, 255), 2)

    return img_color

def sad_matching(img_path, template_path):
    """
    问题57：模板匹配(SAD)
    使用绝对差和进行模板匹配

    参数:
        img_path: 输入图像路径
        template_path: 模板图像路径

    返回:
        匹配结果可视化
    """
    # 读取图像
    img = cv2.imread(img_path, cv2.IMREAD_GRAYSCALE)
    template = cv2.imread(template_path, cv2.IMREAD_GRAYSCALE)
    if img is None or template is None:
        raise ValueError("无法读取图像")

    h, w = template.shape
    H, W = img.shape
    result = np.zeros((H-h+1, W-w+1), dtype=np.float32)

    # 计算SAD
    for y in range(H-h+1):
        for x in range(W-w+1):
            diff = np.abs(img[y:y+h, x:x+w] - template)
            result[y, x] = np.sum(diff)

    # 归一化结果
    result = cv2.normalize(result, None, 0, 255, cv2.NORM_MINMAX).astype(np.uint8)

    # 找到最佳匹配位置
    min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(result)
    top_left = min_loc
    bottom_right = (top_left[0] + w, top_left[1] + h)

    # 在原图上绘制矩形框
    img_color = cv2.cvtColor(img, cv2.COLOR_GRAY2BGR)
    cv2.rectangle(img_color, top_left, bottom_right, (0, 0, 255), 2)

    return img_color

def ncc_matching(img_path, template_path):
    """
    问题58：模板匹配(NCC)
    使用归一化互相关进行模板匹配

    参数:
        img_path: 输入图像路径
        template_path: 模板图像路径

    返回:
        匹配结果可视化
    """
    # 读取图像
    img = cv2.imread(img_path, cv2.IMREAD_GRAYSCALE)
    template = cv2.imread(template_path, cv2.IMREAD_GRAYSCALE)
    if img is None or template is None:
        raise ValueError("无法读取图像")

    h, w = template.shape
    H, W = img.shape
    result = np.zeros((H-h+1, W-w+1), dtype=np.float32)

    # 计算模板的范数
    template_norm = np.sqrt(np.sum(template * template))

    # 计算NCC
    for y in range(H-h+1):
        for x in range(W-w+1):
            window = img[y:y+h, x:x+w]
            window_norm = np.sqrt(np.sum(window * window))
            if window_norm > 0 and template_norm > 0:
                result[y, x] = np.sum(window * template) / (window_norm * template_norm)

    # 归一化结果
    result = cv2.normalize(result, None, 0, 255, cv2.NORM_MINMAX).astype(np.uint8)

    # 找到最佳匹配位置
    min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(result)
    top_left = max_loc  # NCC使用最大值
    bottom_right = (top_left[0] + w, top_left[1] + h)

    # 在原图上绘制矩形框
    img_color = cv2.cvtColor(img, cv2.COLOR_GRAY2BGR)
    cv2.rectangle(img_color, top_left, bottom_right, (0, 0, 255), 2)

    return img_color

def zncc_matching(img_path, template_path):
    """
    问题59：模板匹配(ZNCC)
    使用零均值归一化互相关进行模板匹配

    参数:
        img_path: 输入图像路径
        template_path: 模板图像路径

    返回:
        匹配结果可视化
    """
    # 读取图像
    img = cv2.imread(img_path, cv2.IMREAD_GRAYSCALE)
    template = cv2.imread(template_path, cv2.IMREAD_GRAYSCALE)
    if img is None or template is None:
        raise ValueError("无法读取图像")

    h, w = template.shape
    H, W = img.shape
    result = np.zeros((H-h+1, W-w+1), dtype=np.float32)

    # 计算模板的均值和标准差
    template_mean = np.mean(template)
    template_std = np.std(template)

    # 计算ZNCC
    for y in range(H-h+1):
        for x in range(W-w+1):
            window = img[y:y+h, x:x+w]
            window_mean = np.mean(window)
            window_std = np.std(window)
            if window_std > 0 and template_std > 0:
                zncc = np.sum((window - window_mean) * (template - template_mean)) / (window_std * template_std * h * w)
                result[y, x] = zncc

    # 归一化结果
    result = cv2.normalize(result, None, 0, 255, cv2.NORM_MINMAX).astype(np.uint8)

    # 找到最佳匹配位置
    min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(result)
    top_left = max_loc  # ZNCC使用最大值
    bottom_right = (top_left[0] + w, top_left[1] + h)

    # 在原图上绘制矩形框
    img_color = cv2.cvtColor(img, cv2.COLOR_GRAY2BGR)
    cv2.rectangle(img_color, top_left, bottom_right, (0, 0, 255), 2)

    return img_color

def feature_point_matching(img_path1, img_path2):
    """
    问题60：特征点匹配
    使用特征描述子进行图像匹配

    参数:
        img_path1: 第一张图像路径
        img_path2: 第二张图像路径

    返回:
        匹配结果可视化
    """
    # 读取图像
    img1 = cv2.imread(img_path1)
    img2 = cv2.imread(img_path2)
    if img1 is None or img2 is None:
        raise ValueError("无法读取图像")

    # 转换为灰度图
    gray1 = cv2.cvtColor(img1, cv2.COLOR_BGR2GRAY)
    gray2 = cv2.cvtColor(img2, cv2.COLOR_BGR2GRAY)

    # 创建SIFT检测器
    sift = cv2.SIFT_create()

    # 检测关键点和计算描述子
    keypoints1, descriptors1 = sift.detectAndCompute(gray1, None)
    keypoints2, descriptors2 = sift.detectAndCompute(gray2, None)

    # 创建BF匹配器
    bf = cv2.BFMatcher(cv2.NORM_L2, crossCheck=True)

    # 进行特征匹配
    matches = bf.match(descriptors1, descriptors2)

    # 按距离排序
    matches = sorted(matches, key=lambda x: x.distance)

    # 绘制前10个匹配
    result = cv2.drawMatches(img1, keypoints1, img2, keypoints2, matches[:10], None,
                            flags=cv2.DrawMatchesFlags_NOT_DRAW_SINGLE_POINTS)

    return result

def main(problem_id=56):
    """
    主函数，通过problem_id选择要运行的问题

    参数:
        problem_id: 问题编号，默认为56
    """
    input_path = "../images/imori.jpg"
    template_path = "../images/imori_part.jpg"

    if problem_id == 56:
        result = ssd_matching(input_path, template_path)
        output_path = "../images/answer_56.jpg"
        title = "SSD模板匹配"
    elif problem_id == 57:
        result = sad_matching(input_path, template_path)
        output_path = "../images/answer_57.jpg"
        title = "SAD模板匹配"
    elif problem_id == 58:
        result = ncc_matching(input_path, template_path)
        output_path = "../images/answer_58.jpg"
        title = "NCC模板匹配"
    elif problem_id == 59:
        result = zncc_matching(input_path, template_path)
        output_path = "../images/answer_59.jpg"
        title = "ZNCC模板匹配"
    elif problem_id == 60:
        result = feature_point_matching(input_path, template_path)
        output_path = "../images/answer_60.jpg"
        title = "特征点匹配"
    else:
        print(f"问题 {problem_id} 不存在")
        return

    # 保存结果
    cv2.imwrite(output_path, result)

    # 显示结果
    cv2.imshow(title, result)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

if __name__ == "__main__":
    import sys
    problem_id = int(sys.argv[1]) if len(sys.argv) > 1 else 56
    main(problem_id)

File: cpp/basic/image_matching.cpp
#include <basic/image_matching.hpp>
#include <cmath>

namespace ip101 {

using namespace cv;
using namespace std;

namespace {
// Internal constants
constexpr int CACHE_LINE = 64;    // CPU cache line size (bytes)
constexpr int SIMD_WIDTH = 32;    // AVX2 SIMD vector width (bytes)
constexpr int BLOCK_SIZE = 16;    // Block processing size

// Memory alignment helper function
inline uchar* alignPtr(uchar* ptr, size_t align = CACHE_LINE) {
    return (uchar*)(((size_t)ptr + align - 1) & ~(align - 1));
}

// SIMD optimized version of SSD computation
void compute_ssd_simd(const Mat& src, const Mat& templ, Mat& result) {
    int h = templ.rows;
    int w = templ.cols;
    int H = src.rows;
    int W = src.cols;

    result.create(H-h+1, W-w+1, CV_32F);
    result = Scalar(0);

    #pragma omp parallel for
    for (int y = 0; y < H-h+1; y++) {
        for (int x = 0; x < W-w+1; x++) {
            float sum = 0;
            for (int i = 0; i < h; i++) {
                const uchar* src_ptr = src.ptr<uchar>(y+i) + x;
                const uchar* templ_ptr = templ.ptr<uchar>(i);

                // Use AVX2 for vectorized computation
                for (int j = 0; j < w; j += 8) {
                    __m256i src_vec = _mm256_cvtepu8_epi32(_mm_loadl_epi64((__m128i*)(src_ptr+j)));
                    __m256i templ_vec = _mm256_cvtepu8_epi32(_mm_loadl_epi64((__m128i*)(templ_ptr+j)));
                    __m256i diff = _mm256_sub_epi32(src_vec, templ_vec);
                    __m256i square = _mm256_mullo_epi32(diff, diff);

                    float temp[8];
                    _mm256_storeu_ps(temp, _mm256_cvtepi32_ps(square));
                    for (int k = 0; k < 8 && j+k < w; k++) {
                        sum += temp[k];
                    }
                }
            }
            result.at<float>(y, x) = sum;
        }
    }
}

// SIMD optimized version of SAD computation
void compute_sad_simd(const Mat& src, const Mat& templ, Mat& result) {
    int h = templ.rows;
    int w = templ.cols;
    int H = src.rows;
    int W = src.cols;

    result.create(H-h+1, W-w+1, CV_32F);
    result = Scalar(0);

    #pragma omp parallel for
    for (int y = 0; y < H-h+1; y++) {
        for (int x = 0; x < W-w+1; x++) {
            float sum = 0;
            for (int i = 0; i < h; i++) {
                const uchar* src_ptr = src.ptr<uchar>(y+i) + x;
                const uchar* templ_ptr = templ.ptr<uchar>(i);

                // Use AVX2 for vectorized computation
                for (int j = 0; j < w; j += 8) {
                    __m256i src_vec = _mm256_cvtepu8_epi32(_mm_loadl_epi64((__m128i*)(src_ptr+j)));
                    __m256i templ_vec = _mm256_cvtepu8_epi32(_mm_loadl_epi64((__m128i*)(templ_ptr+j)));
                    __m256i diff = _mm256_abs_epi32(_mm256_sub_epi32(src_vec, templ_vec));

                    float temp[8];
                    _mm256_storeu_ps(temp, _mm256_cvtepi32_ps(diff));
                    for (int k = 0; k < 8 && j+k < w; k++) {
                        sum += temp[k];
                    }
                }
            }
            result.at<float>(y, x) = sum;
        }
    }
}

// SIMD optimized version of NCC computation
void compute_ncc_simd(const Mat& src, const Mat& templ, Mat& result) {
    int h = templ.rows;
    int w = templ.cols;
    int H = src.rows;
    int W = src.cols;

    result.create(H-h+1, W-w+1, CV_32F);
    result = Scalar(0);

    // Calculate template norm
    float templ_norm = 0;
    for (int i = 0; i < h; i++) {
        const uchar* templ_ptr = templ.ptr<uchar>(i);
        for (int j = 0; j < w; j++) {
            templ_norm += templ_ptr[j] * templ_ptr[j];
        }
    }
    templ_norm = sqrt(templ_norm);

    #pragma omp parallel for
    for (int y = 0; y < H-h+1; y++) {
        for (int x = 0; x < W-w+1; x++) {
            float window_norm = 0;
            float dot_product = 0;

            for (int i = 0; i < h; i++) {
                const uchar* src_ptr = src.ptr<uchar>(y+i) + x;
                const uchar* templ_ptr = templ.ptr<uchar>(i);

                // Use AVX2 for vectorized computation
                for (int j = 0; j < w; j += 8) {
                    __m256i src_vec = _mm256_cvtepu8_epi32(_mm_loadl_epi64((__m128i*)(src_ptr+j)));
                    __m256i templ_vec = _mm256_cvtepu8_epi32(_mm_loadl_epi64((__m128i*)(templ_ptr+j)));

                    // Calculate dot product
                    __m256i product = _mm256_mullo_epi32(src_vec, templ_vec);
                    float temp[8];
                    _mm256_storeu_ps(temp, _mm256_cvtepi32_ps(product));
                    for (int k = 0; k < 8 && j+k < w; k++) {
                        dot_product += temp[k];
                    }

                    // Calculate window norm
                    __m256i square = _mm256_mullo_epi32(src_vec, src_vec);
                    _mm256_storeu_ps(temp, _mm256_cvtepi32_ps(square));
                    for (int k = 0; k < 8 && j+k < w; k++) {
                        window_norm += temp[k];
                    }
                }
            }

            window_norm = sqrt(window_norm);
            if (window_norm > 0 && templ_norm > 0) {
                result.at<float>(y, x) = dot_product / (window_norm * templ_norm);
            }
        }
    }
}

// SIMD optimized version of ZNCC computation
void compute_zncc_simd(const Mat& src, const Mat& templ, Mat& result) {
    int h = templ.rows;
    int w = templ.cols;
    int H = src.rows;
    int W = src.cols;

    result.create(H-h+1, W-w+1, CV_32F);
    result = Scalar(0);

    // Calculate template mean and standard deviation
    float templ_mean = 0;
    float templ_std = 0;
    for (int i = 0; i < h; i++) {
        const uchar* templ_ptr = templ.ptr<uchar>(i);
        for (int j = 0; j < w; j++) {
            templ_mean += templ_ptr[j];
        }
    }
    templ_mean /= (h * w);

    for (int i = 0; i < h; i++) {
        const uchar* templ_ptr = templ.ptr<uchar>(i);
        for (int j = 0; j < w; j++) {
            float diff = templ_ptr[j] - templ_mean;
            templ_std += diff * diff;
        }
    }
    templ_std = sqrt(templ_std / (h * w));

    #pragma omp parallel for
    for (int y = 0; y < H-h+1; y++) {
        for (int x = 0; x < W-w+1; x++) {
            // Calculate window mean and standard deviation
            float window_mean = 0;
            float window_std = 0;
            float zncc = 0;

            for (int i = 0; i < h; i++) {
                const uchar* src_ptr = src.ptr<uchar>(y+i) + x;
                for (int j = 0; j < w; j++) {
                    window_mean += src_ptr[j];
                }
            }
            window_mean /= (h * w);

            for (int i = 0; i < h; i++) {
                const uchar* src_ptr = src.ptr<uchar>(y+i) + x;
                const uchar* templ_ptr = templ.ptr<uchar>(i);
                for (int j = 0; j < w; j++) {
                    float src_diff = src_ptr[j] - window_mean;
                    float templ_diff = templ_ptr[j] - templ_mean;
                    window_std += src_diff * src_diff;
                    zncc += src_diff * templ_diff;
                }
            }
            window_std = sqrt(window_std / (h * w));

            if (window_std > 0 && templ_std > 0) {
                result.at<float>(y, x) = zncc / (window_std * templ_std * h * w);
            }
        }
    }
}

} // anonymous namespace

void ssd_matching(const Mat& src, const Mat& templ, Mat& result, int method) {
    CV_Assert(!src.empty() && !templ.empty());
    CV_Assert(src.type() == CV_8UC1 && templ.type() == CV_8UC1);

    compute_ssd_simd(src, templ, result);
}

void sad_matching(const Mat& src, const Mat& templ, Mat& result) {
    CV_Assert(!src.empty() && !templ.empty());
    CV_Assert(src.type() == CV_8UC1 && templ.type() == CV_8UC1);

    compute_sad_simd(src, templ, result);
}

void ncc_matching(const Mat& src, const Mat& templ, Mat& result) {
    CV_Assert(!src.empty() && !templ.empty());
    CV_Assert(src.type() == CV_8UC1 && templ.type() == CV_8UC1);

    compute_ncc_simd(src, templ, result);
}

void zncc_matching(const Mat& src, const Mat& templ, Mat& result) {
    CV_Assert(!src.empty() && !templ.empty());
    CV_Assert(src.type() == CV_8UC1 && templ.type() == CV_8UC1);

    compute_zncc_simd(src, templ, result);
}

void feature_point_matching(const Mat& src1, const Mat& src2,
                          vector<DMatch>& matches,
                          vector<KeyPoint>& keypoints1,
                          vector<KeyPoint>& keypoints2) {
    CV_Assert(!src1.empty() && !src2.empty());
    CV_Assert(src1.type() == CV_8UC1 && src2.type() == CV_8UC1);

    // Use SIFT feature detector and descriptor
    Ptr<Feature2D> sift = SIFT::create();
    Mat descriptors1, descriptors2;

    // Detect keypoints and compute descriptors
    sift->detectAndCompute(src1, noArray(), keypoints1, descriptors1);
    sift->detectAndCompute(src2, noArray(), keypoints2, descriptors2);

    // Use FLANN matcher for feature matching
    Ptr<DescriptorMatcher> matcher = DescriptorMatcher::create(DescriptorMatcher::FLANNBASED);
    vector<vector<DMatch>> knn_matches;
    matcher->knnMatch(descriptors1, descriptors2, knn_matches, 2);

    // Filter good matches using Lowe's ratio test
    const float ratio_thresh = 0.7f;
    matches.clear();
    for (size_t i = 0; i < knn_matches.size(); i++) {
        if (knn_matches[i][0].distance < ratio_thresh * knn_matches[i][1].distance) {
            matches.push_back(knn_matches[i][0]);
        }
    }
}

Mat draw_matching_result(const Mat& src, const Mat& templ, const Mat& result, int method) {
    CV_Assert(!src.empty() && !templ.empty() && !result.empty());

    Mat img_color;
    if (src.channels() == 1) {
        cvtColor(src, img_color, COLOR_GRAY2BGR);
    } else {
        img_color = src.clone();
    }

    double min_val, max_val;
    Point min_loc, max_loc;
    minMaxLoc(result, &min_val, &max_val, &min_loc, &max_loc);

    Point match_loc;
    if (method == TM_SQDIFF || method == TM_SQDIFF_NORMED) {
        match_loc = min_loc;
    } else {
        match_loc = max_loc;
    }

    rectangle(img_color, match_loc,
             Point(match_loc.x + templ.cols, match_loc.y + templ.rows),
             Scalar(0, 0, 255), 2);

    return img_color;
}

} // namespace ip101